### PR TITLE
support refresh region cache

### PR DESCRIPTION
**Pull request by Smityz**
close: https://github.com/tikv/client-go/issues/730

This PR support two functions:
1. support gets all regions' info when NewClient
2. support timely refresh region cache

A small part of the code structure has been refactored, if you have any suggestions, please let me know.

### CODE REVIEW

**Comment by Smityz** [comment]
@disksing Hi master, could you please help me to review this feature?

File: internal/locate/region_cache.go [hunk 1]
```diff
@@ -398,6 +398,33 @@ func (r *Region) isValid() bool {
...
+	r.regions = make(map[RegionVerID]*Region)
+	r.latestVersions = make(map[uint64]RegionVerID)
+	r.sorted = NewSortedRegions(btreeDegree)
+	for _, region := range rs {
+		r.insertRegionToCache(region)
+	}
+	return r
+}
+
+func (mu *regionIndexMu) refresh(btreeDegree int, r []*Region) {
```
Comment (disksing) [inline internal/locate/region_cache.go:419]: I guess we don't need to `btreeDegree` argument as it is a constant that is always 32.

File: internal/locate/region_cache.go [hunk 2]
```diff
@@ -398,6 +398,33 @@ func (r *Region) isValid() bool {
...
+	sorted         *SortedRegions          // cache regions are organized as sorted key to region ref mapping
+}
+
+func newRegionIndexMu(btreeDegree int, rs []*Region) *regionIndexMu {
+	r := &regionIndexMu{}
+	r.regions = make(map[RegionVerID]*Region)
+	r.latestVersions = make(map[uint64]RegionVerID)
+	r.sorted = NewSortedRegions(btreeDegree)
+	for _, region := range rs {
+		r.insertRegionToCache(region)
```
[the remark this answers is not in the document]
Comment (Smityz) [reply internal/locate/region_cache.go:414]: In practical use, this operation is performed in the background and does not affect the B-tree currently in use. Should I add some comments to this function?

File: internal/locate/region_cache.go (hunk 1 above)
Comment (Smityz) [reply internal/locate/region_cache.go:419]: ok, I will modify it later

File: internal/locate/region_cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (disksing) [reply internal/locate/region_cache.go:414]: > In practical use, this operation is performed in the background and does not affect the B-tree currently in use. Should I add some comments to this function?

got it.

**Comment by Smityz** [comment]
updated, please take a look~ @disksing

File: internal/locate/region_cache.go [hunk 3]
```diff
@@ -1602,6 +1636,46 @@ func (c *RegionCache) scanRegionsFromCache(bo *retry.Backoffer, startKey, endKey
...
+		regions, err := c.scanRegions(bo, startKey, nil, 10000)
+		if err != nil {
+			return err
+		}
+		if len(regions) > 0 && regions[len(regions)-1].meta.EndKey == nil {
+			totalRegions = append(totalRegions, regions...)
+			break
+		}
+		totalRegions = append(totalRegions, regions...)
+		startKey = regions[len(regions)-1].meta.EndKey
```
Comment (disksing) [inline internal/locate/region_cache.go:1673]: it should be better like this:

totalRegions = append(totalRegions, regions...)
if len(regions) == 0 || len(regions[len(regions)-1].meta.EndKey) == 0 {
   break
}
startKey = regions[len(regions)-1].meta.EndKey

**Review by disksing — COMMENTED** [review]
the rest looks good to me!

File: internal/locate/region_cache.go (hunk 3 above)
Comment (Smityz) [reply internal/locate/region_cache.go]: updated

**Comment by disksing** [comment]
Thanks @Smityz

**Pull request merged**